### PR TITLE
KAFKA-12947 Replace EasyMock and PowerMock with Mockito for Streams…

### PR DESCRIPTION
**Pull request by wycccccc**
Development of EasyMock and PowerMock has stagnated while Mockito continues to be actively developed. With the new Java cadence, it's a problem to depend on libraries that do bytecode generation and are not actively maintained. In addition, Mockito is also easier to use.[KAFKA-7438](https://issues.apache.org/jira/browse/KAFKA-7438)

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by ijuma** [comment]
Thanks for the PR. Can we please update https://github.com/apache/kafka/blob/trunk/build.gradle#L374 as well?

**Comment by wycccccc** [comment]
@ijuma Thanks for the hint，have updated.

**Comment by ijuma** [comment]
@vvcephei @ableegoldman can one of you review this PR please? It gets us closer to disabling the JDK 15 builds.

**Comment by ableegoldman** [comment]
cc also @cadonna who's familiar with these specific tests I think

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/graph/TableSourceNodeTest.java [hunk 1]
```diff
@@ -23,43 +23,32 @@
...
 
     @Test
     public void shouldConnectStateStoreToInputTopicIfInputTopicIsUsedAsChangelog() {
         final boolean shouldReuseSourceTopicForChangelog = true;
         topologyBuilder.connectSourceStoreAndTopic(STORE_NAME, TOPIC);
-        EasyMock.replay(topologyBuilder);
 
         buildTableSourceNode(shouldReuseSourceTopicForChangelog);
-
-        EasyMock.verify(topologyBuilder);
```
Comment (ableegoldman) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/graph/TableSourceNodeTest.java:50]: Admittedly I don't know much about Mockito, but it looks like we're removing the one verification step that's actually testing something in this test without replacing it. Isn't there a `verify()` for Mockito?

Comment (wycccccc) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/graph/TableSourceNodeTest.java:50]: In my opinion, the `verfiy(object) `in easymock is to verify whether the stubbed method is actually called.
And this step will be automatically detected in mockito. The call form of verfiy method in mockito is `verfit(object).callmethod`, which is used for some specific detections, such as the number of times the method is called (atMostOnec, atLeastOnec) .
So unless it is a specific test, the general easymock verify is not required by mockito.

**Comment by ableegoldman** [comment]
I think this looks good but there's a merge conflict, can you fix that?

**Comment by wycccccc** [comment]
@ableegoldman Have resloved conflict, thanks for your review.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateManagerUtilTest.java [hunk 2]
```diff
@@ -42,77 +39,65 @@
...
     }
 
     @Test
     public void testRegisterStateStoreWhenTopologyEmpty() {
-        expect(topology.stateStores()).andReturn(emptyList());
-
-        ctrl.checkOrder(true);
-        ctrl.replay();
+        when(topology.stateStores()).thenReturn(emptyList());
+        inOrder(topology);
```
Comment (chia7712) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StateManagerUtilTest.java]: there is a single mock method. Does it need to check order?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateManagerUtilTest.java [hunk 3]
```diff
@@ -42,77 +39,65 @@
...
+        when(stateManager.taskId()).thenReturn(taskId);
 
-        expect(stateManager.taskId()).andReturn(taskId);
+        when(stateDirectory.lock(taskId)).thenReturn(false);
 
-        expect(stateDirectory.lock(taskId)).andReturn(false);
-
-        ctrl.checkOrder(true);
-        ctrl.replay();
+        inOrder(topology, stateManager, stateDirectory);
```
Comment (chia7712) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StateManagerUtilTest.java]: you have to check the execution order through the return object from `inOrder`

**Review by chia7712 — COMMENTED** [review]
@wycccccc thanks for this patch. it seems to me `inOrder` is not used correctly. please check it. thanks.

**Comment by ijuma** [comment]
@wycccccc Are you planning to address the outstanding review comments?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateManagerUtilTest.java (hunk 3 above)
Comment (wycccccc) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StateManagerUtilTest.java]: I have gotten the method how to use inorder and the problem has been solved.

**Comment by wycccccc** [comment]
@ijuma Sure,I have resolved the problem.If there are other problems, I will solve them immediately.

**Comment by clolov** [comment]
Hello @ijuma, @ableegoldman, @wycccccc! I have been working on trying to upgrade Kafka Streams from JUnit 4 to JUnit 5 (https://issues.apache.org/jira/browse/KAFKA-7342) and I was wondering what is the state of this pull request and all the ones listed in https://issues.apache.org/jira/browse/KAFKA-7438? Is there something I can help with to get them to completion?

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/graph/TableSourceNodeTest.java (hunk 1 above)
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/graph/TableSourceNodeTest.java:50]: @wycccccc I do not agree. You have to add `verify(topologyBuilder).connecSourceStoreAndTopic(STORE_NAME, TOPIC);`. How should Mockito know what to verify?

**Comment by cadonna** [comment]
@clolov Thank you for your interest and help! Since this PR was not touched more than half a year, I would be fine to close it and you can open a new one.

**Comment by cadonna** [comment]
@wycccccc Do you plan to still work on this PR?

**Comment by clolov** [comment]
Thank you for the reply @cadonna! Okay, I will wait for a week starting today and I will try to reach out to @wycccccc over email as well should I manage to find it across GitHub or JIRA. If I do not get a response I will then open new pull requests.

**Comment by dplavcic** [comment]
> Thank you for the reply @cadonna! Okay, I will wait for a week starting today and I will try to reach out to @wycccccc over email as well should I manage to find it across GitHub or JIRA. If I do not get a response I will then open new pull requests.

@clolov, I see you're working intensively on the [Migrate streams modules to JUnit 5](https://issues.apache.org/jira/browse/KAFKA-7342) (which is great), and I hope you don't mind me jumping in and helping with what was started in the [Replace EasyMock and PowerMock with Mockito](https://issues.apache.org/jira/browse/KAFKA-7438).

I have created a new PR #12449 for this since the initial author is no longer accessible.

**Pull request closed without merging**